helix-core/src/indent.rs

```diff
@@ -459,40 +492,119 @@ fn query_indents(
 ///     },
 /// );
 /// ```
+#[allow(clippy::too_many_arguments)]
 pub fn treesitter_indent_for_pos(
     query: &Query,
     syntax: &Syntax,
     indent_style: &IndentStyle,
+    tab_width: usize,
     text: RopeSlice,
     line: usize,
     pos: usize,
     new_line: bool,
 ) -> Option<String> {
     let byte_pos = text.char_to_byte(pos);
+    // The innermost tree-sitter node which is considered for the indent
+    // computation. It may change if some predeceding node is extended
     let mut node = syntax
         .tree()
         .root_node()
         .descendant_for_byte_range(byte_pos, byte_pos)?;
-    let mut first_in_line = get_first_in_line(node, byte_pos, new_line);
-    let new_line_break = if new_line {
-        Some((line, byte_pos))
-    } else {
-        None
-    };
-    let query_result = crate::syntax::PARSER.with(|ts_parser| {
+    let (query_result, prev_child) = crate::syntax::PARSER.with(|ts_parser| {
         let mut ts_parser = ts_parser.borrow_mut();
         let mut cursor = ts_parser.cursors.pop().unwrap_or_else(QueryCursor::new);
+        // The query range should intersect with all nodes directly preceding
+        // the cursor in case one of them is extended.
+        // prev_child is the deepest such node.
+        let (query_range, prev_child) = {
+            // TODO Is there some way we can reuse this cursor?
+            let mut tree_cursor = node.walk();
+            let mut prev_child = None;
+            for child in node.children(&mut tree_cursor) {
+                if child.byte_range().end <= byte_pos {
+                    prev_child = Some(child);
+                }
+            }
+            match prev_child {
+                Some(mut prev_child) => {
+                    // Get the deepest directly preceding node
+                    while prev_child.child_count() > 0 {
+                        prev_child = prev_child.child(prev_child.child_count() - 1).unwrap();
+                    }
+                    (
+                        prev_child.byte_range().end - 1..byte_pos + 1,
+                        Some(prev_child),
+                    )
+                }
+                None => (byte_pos..byte_pos + 1, None),
+            }
+        };
```

Comment:
I think this block can be simplified a bit - `query_range` can be calculated after `prev_child` is determined with a map+unwrap_or. So the whole `let (query_range, prev_child)` is not needed

Reply:
Agreed. I also noticed that the treesitter parser was borrowed unnecessarily early, so I fixed that too.